File: contents/09/basic-2/solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define REP(i, n) for (int i = 0; i < (n); i++)

char maze[3030][3030];
int dist[3030][3030];
bool visited[3030][3030];
const int INF = 1e9 + 1;
int dx[4] = {0, 0, -1, 1}, dy[4] = {-1, 1, 0, 0};
using pii = pair<int, int>;
void solve()
{
    int H, W;
    cin >> H >> W;
    int sx, sy, gx, gy;
    cin >> sx >> sy >> gx >> gy;
    sx--, sy--, gx--, gy--;
    REP(i, H)
    {
        REP(j, W) { cin >> maze[i][j]; }
    }
    REP(i, H)
    REP(j, W)
    dist[i][j] = INF;
    dist[sx][sy] = 0;

    queue<pii> que;
    que.push({sx, sy});
    visited[sx][sy] = true;
    while (!que.empty())
    {
        pii p = que.front();
        que.pop();
        int x = p.first, y = p.second;
        REP(k, 4)
        {
            int nx = x + dx[k], ny = y + dy[k];
            if (nx >= 0 && nx < H && ny >= 0 && ny < W && !visited[nx][ny] && maze[nx][ny] == '.')
            {
                visited[nx][ny] = true;
                dist[nx][ny] = dist[x][y] + 1;
                que.push({nx, ny});
            }
        }
    }
    int ans = dist[gx][gy];
    assert(ans < INF);
    cout << ans << endl;
}

signed main() { solve(); }

File: contents/09/basic-2/solution.py
from collections import deque
H, W = map(int, input().split())
sy, sx = map(int, input().split())
gy, gx = map(int, input().split())
sy -= 1
sx -= 1
gy -= 1
gx -= 1
A = []*H
for i in range(H):
    s = input()
    A.append(s)

DX = [0, 0, 1, -1]
DY = [1, -1, 0, 0]
INF = 10**9

d = [[INF for j in range(W)]for i in range(H)]
d[sy][sx] = 0

visited = [[False for j in range(W)]for i in range(H)]
visited[sy][sx] = True
que = deque()
que.append((sy, sx))
while que:
    y, x = que.popleft()
    for k in range(4):
        nx = x+DX[k]
        ny = y+DY[k]
        if ny >= 0 and ny < H and nx >= 0 and nx < W and A[ny][nx] != '#' and not visited[ny][nx]:
            visited[ny][nx] = True
            d[ny][nx] = d[y][x]+1
            que.append((ny, nx))

print(d[gy][gx])
